syntax = "proto3";

package ventus.cv;

option cc_enable_arenas = true;

// Scene classification request
message VerifyImageRequest {
    // Raw image bytes (JPEG/PNG encoded)
    bytes image_data = 1;
    
    // Optional: pre-decoded image dimensions
    int32 width = 2;
    int32 height = 3;
    
    // Request ID for tracing
    string request_id = 4;
    
    // Minimum confidence threshold (0.0-1.0)
    float min_confidence = 5;
}

// Individual scene label prediction
message SceneLabel {
    string label = 1;
    float confidence = 2;
}

// Face detection result
message FaceDetection {
    // Bounding box (normalized 0.0-1.0)
    float x = 1;
    float y = 2;
    float width = 3;
    float height = 4;
    float confidence = 5;
}

// Verification response
message VerifyImageResponse {
    // Overall verification result
    bool is_outdoor = 1;
    bool face_detected = 2;
    bool verification_passed = 3;
    
    // Confidence scores
    float outdoor_confidence = 4;
    float face_confidence = 5;
    
    // Detailed predictions
    repeated SceneLabel scene_labels = 6;
    repeated FaceDetection faces = 7;
    
    // Performance metrics
    int64 inference_time_ms = 8;
    int64 preprocessing_time_ms = 9;
    
    // Error handling
    bool success = 10;
    string error_message = 11;
    
    // Request tracing
    string request_id = 12;
}

// Health check
message HealthRequest {}

message HealthResponse {
    bool healthy = 1;
    string version = 2;
    int64 uptime_seconds = 3;
    int32 requests_processed = 4;
}

// Model info
message ModelInfoRequest {}

message ModelInfoResponse {
    string model_name = 1;
    string model_version = 2;
    int32 num_classes = 3;
    repeated string class_labels = 4;
    int32 input_width = 5;
    int32 input_height = 6;
}

// Verification service
service VerificationService {
    // Verify a single image
    rpc VerifyImage(VerifyImageRequest) returns (VerifyImageResponse);
    
    // Stream verification for batch processing
    rpc VerifyImageStream(stream VerifyImageRequest) returns (stream VerifyImageResponse);
    
    // Health check
    rpc CheckHealth(HealthRequest) returns (HealthResponse);
    
    // Get model information
    rpc GetModelInfo(ModelInfoRequest) returns (ModelInfoResponse);
}

